### PR TITLE
Add support for pipeline configuration service

### PR DESCRIPTION
**Pull request by s00500**
- Add configuration extension flags to server 
- Add httpsignatures dependency
- Add http fetching to config fetcher

This PR provides the first version of an external configuration extension. It is globally configured via flags. In the next version the external http endpoint should be possible to be configured in the repo settings via the webui.
#783

The woodpecker server performs a request to a given http endpoint, supplying repo and build info. The request is signed with a httpsignature for the configuration http service to verify the authenticity request. This spec matched the implementation in droneCI

Only a draft for now, error handling and more still needs to be improved

If you have comments about structure or naming of things please let me know

### CODE REVIEW

**Comment by anbraten** [comment]
@s00500 That looks pretty awesome already. One thing I misunderstood is that the original config fetched by WP from the repo is not being send to the endpoint somehow I had that in mind all the time somehow 🙈 So drone and now WP would send the repo and the user model as requestData to the endpoint instead, right?

One thing we have to thing about / test as well, is how multipipelines will work. Currently I guess WP will simply send a single request to the endpoint, but how would the endpoint be able to return multiple configs / pipelines?

**Comment by s00500** [comment]
> @s00500 That looks pretty awesome already. One thing I misunderstood is that the original config fetched by WP from the repo is not being send to the endpoint somehow I had that in mind all the time somehow 🙈 So drone and now WP would send the repo and the user model as requestData to the endpoint instead, right?
> 
> One thing we have to thing about / test as well, is how multipipelines will work. Currently I guess WP will simply send a single request to the endpoint, but how would the endpoint be able to return multiple configs / pipelines?

Simple, we just make the data response an array. That would handle multipipelines just fine, as before returning 204 would run the configs from the repositories like before

**Comment by 6543** [comment]
If i understand it right you like to be able to load the pipeline config from an external source?

**Comment by s00500** [comment]
@6543 Exactly, I am basically replicating the behavior used here: https://docs.drone.digitalstacks.net/l/en/extensions/configuration-extension

**Comment by 6543** [comment]
I'd like to propose a similar concept as done with the "clone" step. an optional "pipeline config compile" step - and there you could use any plugin - and so add starlarc support as plugin that transparently convert ... github pipeline config etc ...

**Comment by 6543** [comment]
your aproach is fare more simpler ... so it's easyer to implement ... - but does it limit us later in any way?

File: server/plugins/configuration/configuration.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/99designs/httpsignatures-go"
```
Comment (6543) [inline server/plugins/configuration/configuration.go:13]:
```suggestion
	"github.com/99designs/httpsignatures-go"

```

File: server/shared/configFetcher.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 	"time"
 
 	"github.com/rs/zerolog/log"
+	"github.com/woodpecker-ci/woodpecker/server/plugins/configuration"
```
Comment (6543) [inline server/shared/configFetcher.go:11]: sort import

File: server/shared/configFetcher.go [hunk 3]
```diff
@@ -60,6 +63,22 @@ func (cf *configFetcher) fetch(c context.Context, timeout time.Duration, config
 	ctx, cancel := context.WithTimeout(c, timeout)
 	defer cancel()
 
+	if cf.configAPI.IsConfigured() {
+		log.Trace().Msgf("ConfigFetch[%s]: getting config from external http service '%s'", cf.repo.FullName)
```
Comment (6543) [inline server/shared/configFetcher.go:67]: since drone and woodpecker pipeline config differ anyway ...
I think we can add multi pipeline support right away & we should also send the origin config from repo (if exist) to that endpoint

**Comment by anbraten** [comment]
> But finally this is your project, so your call if this makes sense :-) It is definitely straight forward to add.

I really like your current approach, sending the config data as string as another parameter (next to build and repo) would be super cool. In general I think this approach is awesome, because it is so simple that we can easily adjust / improve it in the feature.

File: server/shared/configFetcher.go (hunk 2 above)
Comment (s00500) [reply server/shared/configFetcher.go:11]: what tooling / guideline are you using for the import sorting ? I usually use vscode with the languageserver as default setup...

File: server/shared/configFetcher.go (hunk 3 above)
Comment (s00500) [reply server/shared/configFetcher.go:67]: Could not agree more :-)

**Comment by s00500** [comment]
Ok I added the config sending, multipipeline, some basic testing. Still missing docs

File: server/plugins/configuration/configuration.go (hunk 1 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go:13]: done

File: server/shared/configFetcher.go (hunk 2 above)
Comment (s00500) [reply server/shared/configFetcher.go:11]: Ok I think I got it, main point is the grouping into std / external / woodpecker right ?

Comment (anbraten) [reply server/shared/configFetcher.go:11]: I am using VS-Code as well. For import sorting we use golangci-lint, if that's what you were asking for.

Comment (s00500) [reply server/shared/configFetcher.go:11]: Yep, I have that setup as linting tool too :-)

File: cmd/server/flags.go [hunk 4]
```diff
@@ -161,6 +161,16 @@ var flags = []cli.Flag{
...
+	&cli.StringFlag{
+		EnvVars: []string{"WOODPECKER_YAML_ENDPOINT"},
+		Name:    "configuration-service",
+		Usage:   "configuration service endpoint",
+	},
+	&cli.StringFlag{
+		EnvVars: []string{"WOODPECKER_YAML_SECRET"},
+		Name:    "configuration-service-secret",
+		Usage:   "configuration service secret",
+	},
```
Comment (wxiaoguang) [inline cmd/server/flags.go:173]: The `EnvVars` and `Name` are different, could using the similar names be better?

Most other EnvVars and Names are using similar names

Comment (s00500) [reply cmd/server/flags.go:173]: @wxiaoguang do you have any recommendations ?
Eg:
WOODPECKER_CONFIG_SERVICE_ENDPOINT
WOODPECKER_CONFIG_SERVICE_SECRET

?

Comment (wxiaoguang) [reply cmd/server/flags.go:173]: WOODPECKER_CONFIG_SERVICE_ENDPOINT <=> --config-service-endpoint
WOODPECKER_CONFIG_SERVICE_SECRET <=> --config-service-secret

It sounds very accurate 😊

File: server/api/build.go [hunk 5]
```diff
@@ -453,18 +453,38 @@ func PostBuild(c *gin.Context) {
 		}
 	}
 
-	// fetch the pipeline config from database
+	var yamls []*remote.FileMeta
```
Comment (anbraten) [inline server/api/build.go:456]: What do you think about removing the term `yaml` here. (I know it was there before, so if you don't want to change the whole code here we can also do it in a separate PR)

File: server/plugins/configuration/configuration.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+	"github.com/woodpecker-ci/woodpecker/server/model"
+	"github.com/woodpecker-ci/woodpecker/server/remote"
+)
+
+type ConfigAPI struct {
```
Comment (anbraten) [inline server/plugins/configuration/configuration.go:19]:
```suggestion
type ConfigService struct {
```

File: server/plugins/configuration/configuration.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	return cp.endpoint != ""
+}
+
+func (cp *ConfigAPI) FetchExternalConfig(ctx context.Context, repo *model.Repo, build *model.Build, currentConfig []*remote.FileMeta) (configData []*remote.FileMeta, useOld bool, err error) {
+	// Create request, needs repo
+	response := struct {
+		Pipelines []config `json:"pipelines"`
+	}{}
+
+	currentYamls := make([]*config, len(currentConfig))
```
Comment (anbraten) [inline server/plugins/configuration/configuration.go:50]: You could try to use `_currentConfig` to not have to rename the var:

 ```suggestion
	currentConfig := make([]*config, len(_currentConfig))
```

File: server/plugins/configuration/configuration.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	for i, pipe := range currentConfig {
+		currentYamls[i] = &config{Name: pipe.Name, Data: string(pipe.Data)}
+	}
+
+	status, err := sendRequest(ctx, "POST", cp.endpoint, cp.secret, requestStructure{Repo: repo, Build: build, Configuration: currentYamls}, &response)
```
Comment (anbraten) [inline server/plugins/configuration/configuration.go:55]: Could we change `sendRequest` to not be call by ref on `&response`.

File: server/plugins/configuration/configuration.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+func (cp *ConfigAPI) IsConfigured() bool {
+	return cp.endpoint != ""
+}
+
+func (cp *ConfigAPI) FetchExternalConfig(ctx context.Context, repo *model.Repo, build *model.Build, currentConfig []*remote.FileMeta) (configData []*remote.FileMeta, useOld bool, err error) {
+	// Create request, needs repo
+	response := struct {
+		Pipelines []config `json:"pipelines"`
```
Comment (anbraten) [inline server/plugins/configuration/configuration.go:47]: We should name it pipeline or config 😉 

```suggestion
		Pipelines []config `json:"pipelines"`
```

**Review by anbraten — COMMENTED** [review]
Really like this feature and your Implementation. Just some renaming suggestions.

File: server/api/build.go (hunk 5 above)
Comment (s00500) [reply server/api/build.go:456]: -> F2 Rename Symbol :-)

Will do

File: server/plugins/configuration/configuration.go (hunk 7 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go:50]: Not a super big fan of _ prefix, but I can still get rid of the yaml term :-)

File: server/plugins/configuration/configuration.go (hunk 9 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go:47]: So should I change the field name to Configs ? seems to be more in line with what I did so far

File: server/plugins/configuration/configuration.go (hunk 8 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go:55]: Hm, I took the function from some other part of the codebase, it is mainly done so that the json decoder does use the correct type.... I ca make it explicit defined if you want

File: server/plugins/configuration/configuration.go (hunk 9 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go]: changed

File: server/plugins/configuration/configuration.go (hunk 8 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go]: changed

File: server/plugins/configuration/configuration.go (hunk 7 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go]: changed

File: server/plugins/configuration/configuration.go (hunk 6 above)
Comment (s00500) [reply server/plugins/configuration/configuration.go]: changed

File: server/api/build.go (hunk 5 above)
Comment (s00500) [reply server/api/build.go]: changed

File: docs/docs/30-administration/100-external-configuration-api.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
+# External Configuration API
+
+To provide additional management and preprocessing capabilities for pipeline configuration Woodpecker supports an API for external configuration.
```
[the remark this answers is not in the document]
Comment (s00500) [reply docs/docs/30-administration/100-external-configuration-api.md]: changed

File: server/shared/configFetcher_test.go [hunk 11]
```diff
@@ -63,7 +70,7 @@ func TestFetch(t *testing.T) {
 			expectedError: false,
 		},
 		{
-			name:       "Default config - .drone.yml",
+			name:       "Use old config by 204 response",
```
[the remark this answers is not in the document]
Comment (s00500) [reply server/shared/configFetcher_test.go:73]: changed

File: cmd/server/flags.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (s00500) [reply cmd/server/flags.go]: changed

File: docs/docs/30-administration/100-external-configuration-api.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+Before the run or restart of any pipeline Woodpecker will make a POST request to an external REST API with the current repo and build information, and the current config file. The external API can then send back new pipeline configurations that will be used immediately or responsnd with `HTTP 204` to tell the system to use the existing configuration.
```
[the remark this answers is not in the document]
Comment (s00500) [reply docs/docs/30-administration/100-external-configuration-api.md]: changed

File: docs/docs/30-administration/100-external-configuration-api.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+Every request sent by Woodpecker is signed using a http-signature created with a provided secret. This way the external api can verify the authenticity request from the Woodpecker Instance
```
[the remark this answers is not in the document]
Comment (s00500) [reply docs/docs/30-administration/100-external-configuration-api.md:6]: changed

File: docs/docs/30-administration/100-external-configuration-api.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+```json
+{
+  "pipelines": [
+    {
+      "name": "central-override",
+      "data": "pipeline:\n  backend:\n    image: alpine\n    commands:\n      - echo \"Hello there from ConfigAPI\"\n"
+    }
+  ]
+}
+```
```
[the remark this answers is not in the document]
Comment (s00500) [reply docs/docs/30-administration/100-external-configuration-api.md:104]: Right I was thinking about that, I will setup a repo for this, maybe that makes sense to then also move to the woodpecker org ?

**Comment by s00500** [comment]
Created a simple example, not yet tested: https://github.com/s00500/woodpecker-config-service

File: docs/docs/30-administration/100-external-configuration-api.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
+# External Configuration API
+
+To provide additional management and preprocessing capabilities for pipeline configurations Woodpecker supports an HTTP api which can be enabled to call an external config service.
+Before the run or restart of any pipeline Woodpecker will make a POST request to an external HTTP api sending the current repository, build information and all current config files retrieved from the repository. The external api can then send back new pipeline configurations that will be used immediately or respond with `HTTP 204` to tell the system to use the existing configuration.
+
+Every request sent by Woodpecker is signed using a http-signature using the provided secret from `WOODPECKER_CONFIG_SERVICE_SECRET`. This way the external api can verify the authenticity request from the Woodpecker instance.
+
+A simplistic example configuration service can be found here: [https://github.com/s00500/woodpecker-config-service](https://github.com/s00500/woodpecker-config-service)
```
Comment (anbraten) [inline docs/docs/30-administration/100-external-configuration-api.md:8]: @s00500 I have created a repo inside the woodpecker org and added you as maintainer.

```suggestion
A simplistic example configuration service can be found here: [https://github.com/woodpecker-ci/example-config-service](https://github.com/woodpecker-ci/example-config-service)
```

File: server/shared/configFetcher.go [hunk 16]
```diff
@@ -18,37 +19,56 @@ type ConfigFetcher interface {
...
 type configFetcher struct {
-	remote remote.Remote
-	user   *model.User
-	repo   *model.Repo
-	build  *model.Build
+	remote    remote.Remote
+	user      *model.User
+	repo      *model.Repo
+	build     *model.Build
+	configAPI configuration.ConfigService
```
Comment (anbraten) [inline server/shared/configFetcher.go:26]:
```suggestion
	configService configuration.ConfigService
```

File: server/shared/configFetcher_test.go [hunk 17]
```diff
@@ -237,6 +244,201 @@ func TestFetch(t *testing.T) {
...
+			matchingFiles := make([]string, len(files))
+			for i := range files {
+				matchingFiles[i] = files[i].Name
+			}
+			assert.ElementsMatch(t, tt.expectedFileNames, matchingFiles, "expected some other pipeline files")
+		})
+	}
+}
+
+func TestFetchFromConfigurationAPI(t *testing.T) {
```
Comment (anbraten) [inline server/shared/configFetcher_test.go:268]:
```suggestion
func TestFetchFromConfigService(t *testing.T) {
```

File: server/api/build.go [hunk 18]
```diff
@@ -453,16 +453,36 @@ func PostBuild(c *gin.Context) {
...
+	}
+
+	// If config extension is active we should refetch the config in case something changed
+	if server.Config.Services.ConfigService.IsConfigured() {
+		currentYamls := make([]*remote.FileMeta, len(configs))
+		for i, cfg := range configs {
+			currentYamls[i] = &remote.FileMeta{Name: cfg.Name, Data: cfg.Data}
+		}
+
+		newConfig, useOld, err := server.Config.Services.ConfigService.FetchExternalConfig(c, repo, build, currentYamls)
```
Comment (anbraten) [inline server/api/build.go:477]:
```suggestion
		currentFileMeta := make([]*remote.FileMeta, len(configs))
		for i, cfg := range configs {
			currentFileMeta[i] = &remote.FileMeta{Name: cfg.Name, Data: cfg.Data}
		}

		newConfig, useOld, err := server.Config.Services.ConfigService.FetchExternalConfig(c, repo, build, currentFileMeta)
```

File: server/api/hook.go [hunk 19]
```diff
@@ -191,7 +191,7 @@ func PostHook(c *gin.Context) {
 		msg := "failure to parse yaml from hook"
 		log.Debug().Err(err).Str("repo", repo.FullName).Msg(msg)
 		c.String(http.StatusBadRequest, msg)
-		return
+		// return
```
Comment (anbraten) [inline server/api/hook.go:194]: Is this change related to #811?

Comment (s00500) [reply server/api/hook.go:194]: Huch, did not mean to commit that here, but yes this would be my proposed solution to #811, that should be all it takes

Comment (s00500) [reply server/api/hook.go:194]: Shall I revert this ?

File: server/shared/configFetcher.go (hunk 16 above)
Comment (s00500) [reply server/shared/configFetcher.go]: ✅

File: docs/docs/30-administration/100-external-configuration-api.md (hunk 15 above)
Comment (s00500) [reply docs/docs/30-administration/100-external-configuration-api.md]: ✅ 
That's great, I adapted the docs, will do a final check on it and fix potential bugs

File: server/shared/configFetcher_test.go (hunk 17 above)
Comment (s00500) [reply server/shared/configFetcher_test.go]: ✅

File: server/api/build.go (hunk 18 above)
Comment (s00500) [reply server/api/build.go]: ✅

**Comment by s00500** [comment]
Example service is updated and tested now, also added some more notes to the readme over there :-)

File: server/api/hook.go (hunk 19 above)
Comment (anbraten) [reply server/api/hook.go:194]: Yeah, lets move it into a separate PR.

Comment (s00500) [reply server/api/hook.go]: ✅

File: cmd/server/flags.go [hunk 20]
```diff
@@ -161,6 +161,16 @@ var flags = []cli.Flag{
 		Name:    "gating-service",
 		Usage:   "gated build endpoint",
 	},
+	&cli.StringFlag{
+		EnvVars: []string{"WOODPECKER_CONFIG_SERVICE_ENDPOINT"},
+		Name:    "config-service-endpoint",
+		Usage:   "configuration service endpoint",
```
Comment (anbraten) [inline cmd/server/flags.go:167]:
```suggestion
		Usage:   "url used for calling configuration service endpoint",
```

**Review by anbraten — APPROVED** [review]

**Comment by s00500** [comment]
Awesome, thanks a lot =D
